### PR TITLE
Fix SelectNext focus inconsistencies when mixing mouse and keyboard interaction

### PR DESCRIPTION
**Pull request by acelaya**
Closes https://github.com/hypothesis/frontend-shared/issues/1285

This PR addresses the two problems left in `SelectNext`, both related with option focusing.

* Now, the focus ring is visible for the first option in the listbox when open. The option was previously focused already, but the style was applied to `:focus-visible` and not `:focus`, which prevents the UA from applying the styles. See first example in https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible#examples
* When an option is clicked, we now make sure it is focused just before closing the listbox, becoming the last focused element. This ensures it is focused when the listbox is opened again, regardless of how the last option was selected (via mouse or keyboard).

In order to fix the first issue, options are no longer `Button` components, but simple `li` where we have to override less styles and we can apply focus rings to `:focus` instead of `:focus-visible`.

This also means the listbox is now a `ul` instead of a `div`.

### CODE REVIEW

File: src/components/input/Button.tsx [hunk 1]
```diff
@@ -96,8 +103,11 @@ export default function Button({
       {...htmlAttributes}
       className={classnames(
         {
-          'focus-visible-ring transition-colors whitespace-nowrap flex items-center':
-            styled,
+          'transition-colors whitespace-nowrap flex items-center': styled,
+        },
+        styled && {
+          'focus-visible-ring': focusRing === 'focus-visible',
+          'focus:ring outline-none': focusRing === 'focus',
```
Comment (acelaya) [inline src/components/input/Button.tsx:110]: I don't love having to add a new property to the `Button`, but it's the only way to not affect other use cases.

Basically, the custom `focus-visible-ring` aims to style focused elements based on `:focus-visible`, instead of `:focus`.

Those two are equivalent for editable elements (like inputs or elements with `contenteditable`), but not for `buttons`.

A clicked button becomes focused, but does not become `focus-visible`, preventing the focus ring to render (See https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible for details and examples).

This effectively means that, if the `SelectNext` toggle is clicked, the listbox is opened, the first option is focused, but it does not render a focus ring, causing the second option to render a focus ring when pressing <kbd>ArrowDown</kbd> for the first time.

By using tailwind's standard `focus:ring`, we apply the ring to `:focus`, making sure the first option renders a focus ring when the listbox is opened via mouse, and then we can interact with the list via arrow keys with an intuitive behavior.

File: src/hooks/use-arrow-key-navigation.ts [hunk 2]
```diff
@@ -94,6 +94,10 @@ export function useArrowKeyNavigation(
   const lastFocusedItem = useRef<HTMLElement | null>(null);
 
   useEffect(() => {
+    if (!containerVisible) {
+      return () => {};
+    }
```
Comment (acelaya) [inline src/hooks/use-arrow-key-navigation.ts:99]: This is not strictly related with the other changes on this PR, but I realized we can safely skip any action when `containerVisible` is false, making the rest of the logic easier to reason about.

File: src/hooks/use-arrow-key-navigation.ts [hunk 3]
```diff
@@ -215,10 +215,11 @@ export function useArrowKeyNavigation(
         lastFocusedItem.current.focus();
         return;
       }
+
       const elements = getNavigableElements();
       const targetIndex = elements.indexOf(event.target as HTMLElement);
       if (targetIndex >= 0) {
-        updateTabIndexes(elements, targetIndex);
+        updateTabIndexes(elements, targetIndex, autofocus);
```
Comment (acelaya) [inline src/hooks/use-arrow-key-navigation.ts:222]: This is what was causing the options to not me set as `lastFocusedItem` when being clicked instead of selected with the keyboard.

File: src/components/input/Button.tsx [hunk 4]
```diff
@@ -27,6 +27,12 @@ type ComponentProps = {
    */
   title?: string;
 
+  /**
+   * Whether the focus ring should be applied to `:focus` or `:focus-visible`.
+   * Defaults to focus-visible.
+   */
+  focusRing?: 'focus' | 'focus-visible';
```
Comment (acelaya) [inline src/components/input/Button.tsx:34]: > See https://github.com/hypothesis/frontend-shared/pull/1292/files#r1345926735 for more details on why this has been introduced

I think `focusRing` sounds like a boolean property. Perhaps a better name would be one of `focusRingMode`, `focusRingType` or `focusRingVariation`.

File: src/hooks/test/use-arrow-key-navigation-test.js [hunk 5]
```diff
@@ -30,15 +30,14 @@ function Toolbar({ navigationOptions = {} }) {
...
-  const toggleVisible = useCallback(() => setVisible(prev => !prev), []);
   const options = useMemo(
     () => ({ ...navigationOptions, containerVisible: visible }),
     [navigationOptions, visible],
   );
 
   return (
     <>
-      <button data-testid="toggle" onClick={toggleVisible}>
+      <button data-testid="toggle" onClick={() => setVisible(!visible)}>
```
Comment (acelaya) [inline src/hooks/test/use-arrow-key-navigation-test.js:40]: This was an attempt to fix an uncovered line, but didn't help. I'll remove it.

**Comment by acelaya** [comment]
There's a line not covered, which should be when `containerVisible` changes and the effect is re-triggered. However, the test is not changing this and I'm trying to figure out why.

I'll put this in draft state for now, just in case this ends up having deeper implications.

**Comment by acelaya** [comment]
> There's a line not covered, which should be when `containerVisible` changes and the effect is re-triggered. However, the test is not changing this and I'm trying to figure out why.
> 
> I'll put this in draft state for now, just in case this ends up having deeper implications.

Ok, I was missing some `act` calls to make sure side effects were triggered when clicking a button.

File: src/components/input/Button.tsx (hunk 1 above)
Comment (robertknight) [reply src/components/input/Button.tsx:110]: If we're using a `Button` but overriding basically all the key aspects of it (style, semantics, behavior), then we should consider not using `Button` at all here. The complexity added for non-buttony use cases is going to make it increasingly difficult to maintain this.

FWIW the Headless UI and React Spectrum Picker examples we've been referencing don't use buttons.

Comment (acelaya) [reply src/components/input/Button.tsx:110]: I knew you were going to suggest this 😅

Yeah, I'll do the changes. At this point it probably makes sense to use something else

File: src/components/input/SelectNext.tsx [hunk 6]
```diff
@@ -46,15 +46,19 @@ function SelectOption<T>({
...
   return (
-    <Button
-      variant="custom"
-      classes={classnames(
-        'w-full ring-inset rounded-none !p-0',
-        'border-t first:border-t-0 bg-transparent',
+    <li
+      className={classnames(
+        'w-full ring-inset focus:ring outline-none rounded-none',
+        'border-t first:border-t-0 transition-colors whitespace-nowrap',
```
Comment (acelaya) [inline src/components/input/SelectNext.tsx:52]: Moving for `Button` to `li` allows for less style overrides to be needed, and `focus:ring` to be used instead of `focus-visible-ring`

File: src/components/input/SelectNext.tsx [hunk 7]
```diff
@@ -46,15 +46,19 @@ function SelectOption<T>({
...
         { 'hover:bg-grey-1': !disabled },
         classes,
       )}
       onClick={() => selectValue(value)}
+      onKeyPress={e => {
+        if (e.key === 'Enter') {
+          selectValue(value);
+        }
+      }}
```
Comment (acelaya) [inline src/components/input/SelectNext.tsx:61]: This was transparently handled by `Buttons`, but needs to be explicitly added for `li`s.

File: src/components/input/SelectNext.tsx (hunk 6 above)
Comment (robertknight) [reply src/components/input/SelectNext.tsx:52]: One other style that came for free with buttons was setting `cursor: pointer` (or the equivalent TW style, that will need to be added here too).

File: src/components/input/SelectNext.tsx [hunk 8]
```diff
@@ -46,15 +46,19 @@ function SelectOption<T>({
...
+    <li
+      className={classnames(
+        'w-full ring-inset focus:ring outline-none rounded-none',
+        'border-t first:border-t-0 transition-colors whitespace-nowrap',
         { 'hover:bg-grey-1': !disabled },
         classes,
       )}
       onClick={() => selectValue(value)}
+      onKeyPress={e => {
+        if (e.key === 'Enter') {
```
Comment (robertknight) [inline src/components/input/SelectNext.tsx:58]: Also the Space key should select the item.

**Review by robertknight — COMMENTED** [review]
Just two minor things I spotted with the change to `<li>`:

- Space should select the item as well as Enter (I'm not sure if there is a more general way of getting the keys which have activation behavior)
- The cursor should be set to `pointer`

**Review by robertknight — APPROVED** [review]
